### PR TITLE
feat(controller): Panic or error on mis-matched resource version

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

See #3862

### CODE REVIEW

File: workflow/controller/operator_metrics_test.go [hunk 1]
```diff
@@ -267,21 +269,14 @@ spec:
     name: whalesay-template
     outputs: {}
     retryStrategy:
-      limit: 2
-status:
-  phase: Running
-  startedAt: "2020-07-14T16:26:21Z"
+      limit: 2"
```
Comment (sarabala1979) [inline workflow/controller/operator_metrics_test.go:272]: " is missing

File: workflow/controller/dag_test.go [hunk 2]
```diff
@@ -614,6 +617,7 @@ func TestDagAssessPhaseContinueOnExpandedTaskVariables(t *testing.T) {
 	woc := newWorkflowOperationCtx(wf, controller)
 
 	woc.operate()
+	woc = newWorkflowOperationCtx(woc.wf, controller)
```
Comment (sarabala1979) [inline workflow/controller/dag_test.go:620]: why do we need to create a new context here? we can just use woc.operate() which will have updated wf.

**Review by sarabala1979 — APPROVED** [review]
LGMT

**Comment by alexec** [comment]
```
2020-09-14T19:29:03.8397834Z --- FAIL: TestArgoServerSuite (38.73s)
2020-09-14T19:29:03.8398765Z     --- FAIL: TestArgoServerSuite/TestWorkflowService (2.28s)
2020-09-14T19:29:03.8399892Z         --- FAIL: TestArgoServerSuite/TestWorkflowService/Terminate (2.00s)
2020-09-14T19:29:03.8402489Z             reporter.go:39: 
2020-09-14T19:29:03.8403200Z                 	Error Trace:	reporter.go:39
2020-09-14T19:29:03.8403645Z                 	            				chain.go:21
2020-09-14T19:29:03.8404038Z                 	            				helpers.go:33
2020-09-14T19:29:03.8404430Z                 	            				value.go:81
2020-09-14T19:29:03.8404854Z                 	            				argo_server_test.go:805
2020-09-14T19:29:03.8405261Z                 	            				suite.go:77
2020-09-14T19:29:03.8406033Z                 	Error:      	child 'message' not found in JSON object at 10
2020-09-14T19:29:03.8406853Z                 	Test:       	TestArgoServerSuite/TestWorkflowService/Terminate
2020-09-14T19:29:03.8408024Z         --- FAIL: TestArgoServerSuite/TestWorkflowService/Resubmit (0.00s)
2020-09-14T19:29:03.8408815Z             reporter.go:39: 
2020-09-14T19:29:03.8409252Z                 	Error Trace:	reporter.go:39
2020-09-14T19:29:03.8409685Z                 	            				chain.go:21
2020-09-14T19:29:03.8410203Z                 	            				response.go:585
2020-09-14T19:29:03.8410798Z                 	            				response.go:151
2020-09-14T19:29:03.8411252Z                 	            				argo_server_test.go:813
2020-09-14T19:29:03.8411653Z                 	            				suite.go:77
2020-09-14T19:29:03.8411991Z                 	Error:      	
2020-09-14T19:29:03.8412408Z                 	            	expected status equal to:
2020-09-14T19:29:03.8412789Z                 	            	 "200 OK"
2020-09-14T19:29:03.8413079Z                 	            	
2020-09-14T19:29:03.8413392Z                 	            	but got:
2020-09-14T19:29:03.8413813Z                 	            	 "500 Internal Server Error"
```

**Pull request merged**